leshan-core-cf/src/main/java/org/eclipse/leshan/core/californium/Lwm2mClientEndpointContextMatcher.java

```diff
@@ -0,0 +1,71 @@
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial implementation
+ ******************************************************************************/
+package org.eclipse.leshan.core.californium;
+
+import org.eclipse.californium.elements.auth.X509CertPath;
+import org.eclipse.leshan.core.request.Identity;
+
+import javax.security.auth.x500.X500Principal;
+import java.security.Principal;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.List;
+
+/**
+ * LWM2M principal based endpoint context matcher for LWM2M Clients.
+ * 
+ * Matches DTLS based on the used principal. Requires unique credentials.
+ * 
+ * There is separation between LWM2M Client and Server side EndpointContextMatchers as when client is performing
+ * initial connection it doesn't know real peer identity due to flexibility of certificate usage configuration.
+ *
+ * With certificate usage the server identity may be the intermediate CA or root CA. For this reason the EndpointContext
+ * matcher is relaxed to allow matching to other parts of the chain. This can be done as before the context matcher
+ * there is DTLS Handshake which has verified that certificate usage and chain is expected.
+ *
+ * When server is receiving DTLS connection it always knows the whole peer identity as DTLS handshake is performed first
+ * before any message is being sent to client and then EndpointContext is built with real client information. In server
+ * more lightweight and strict matching can be done there and there is no need to inspect the chain.
+ * 
+ * For more detailed about why this is needed, see
+ * <a href="https://github.com/eclipse/leshan/wiki/LWM2M-Observe#for-dtls">LWM2M-Observe wiki page</a>
+ */
+public class Lwm2mClientEndpointContextMatcher extends Lwm2mEndpointContextMatcher {
+
+    public Lwm2mClientEndpointContextMatcher() {
+    }
+
+    @Override
+    public String getName() {
+        return "lwm2m client correlation";
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * LWM2M x.509 principals has already been checked during TLS handshake.
+     */
+    @Override
+    protected boolean matchPrincipals(Principal requestedPrincipal, Principal availablePrincipal) {
+        if (requestedPrincipal instanceof X500Principal && availablePrincipal instanceof X509CertPath) {
+            // Note: DTLS layer has already verified that certificate usage matches peer and that certificates are valid
+            return true;
+        } else {
+            // Handle non X.509 identities in common LWM2M endpoint context matcher
+            return super.matchPrincipals(requestedPrincipal, availablePrincipal);
+        }
```

Comment:
```suggestion
  return true
```

As I think this is enough. (as explained in https://github.com/eclipse/leshan/pull/903#issuecomment-706158960)
I tested and test pass on my side. (as long as this modification is only apply at client side)

Reply:
Simplified as suggested.